Adding support for using client_type API for interacting with EC2 and support for using filters

VijayantSoni opened the pull request:
1. Adding support for client_type API for accessing EC2 in EC2Hook
2. Adding support for using filters when accessing EC2s (only for client_type API)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](

VijayantSoni committed:
Changing from service api to client api, adding support for accesing ec1s using filters

ashb commented on airflow/providers/amazon/aws/hooks/ec2.py:
```diff
@@ -33,21 +33,90 @@ class EC2Hook(AwsBaseHook):
+    INSTANCES = 'Instances'
+    STATE = 'State'
+    NAME = 'Name'
+    INSTANCE_ID = 'InstanceId'
```
Please don't do this. Use the string directly. Having this extracted out just makes it harder to understand the code at the call site.

VijayantSoni replied:
Agreed Ash, but wanting to understand, is this an anti-pattern in python or against airflow coding convention, because I do see similar being used within the code base.
https://github.com/apache/airflow/blob/639972d995d848b16a3f283576efdbde28b8fdef/airflow/providers/google/cloud/hooks/dataflow.py#L93-L113

VijayantSoni replied:
Holding on this change unless must as this style is already used in the code base.

feluelle replied:
@VijayantSoni in the example you posted those values are outside of the hook. Those are specific _types_.

I agree with Ash on using the strings directly (in your case) as those doesn't really have a category to specify them under, do they?

In the example the category is JobTypes and JobStatus.

VijayantSoni replied:
Thanks @feluelle for the clarification. got it now.

ashb commented on airflow/providers/amazon/aws/hooks/ec2.py:
```diff
@@ -33,21 +33,90 @@ class EC2Hook(AwsBaseHook):
-        super().__init__(resource_type="ec2", *args, **kwargs)
+        super().__init__(client_type="ec2", *args, **kwargs)
 
-    def get_instance(self, instance_id: str):
```
This is only acceptible since the Ec2Hook hasn't yet been included in a release -- otherwise this would be a breaking change and you would have have to include a compatibility layer.

VijayantSoni replied:
missed this, please let me know if this needs to be taken care off or we are good for now.

feluelle replied:
_Unfortunately_ the ec2 integration was released. So we would need take care of it like ash mentioned.

VijayantSoni replied:
👍

ashb commented on airflow/providers/amazon/aws/hooks/ec2.py:
```diff
@@ -33,21 +33,90 @@ class EC2Hook(AwsBaseHook):
+
+        return [
+            instance
+            for reservation in description[self.RESERVATIONS] for instance in reservation[self.INSTANCES]
```
What about spot instances? I think this would break if the query returned any spot instances.

VijayantSoni replied:
As per [boto3 ec2 describe instance api docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.describe_instances) the response does not change in case of spot instances and `InstanceLifecycle` key indicates whether it's an spot instance or not.
```
InstanceLifecycle (string) --

Indicates whether this is a Spot Instance or a Scheduled Instance.
```

ashb commented on airflow/providers/amazon/aws/hooks/ec2.py:
```diff
@@ -58,7 +127,7 @@ def get_instance_state(self, instance_id: str) -> str:
         :rtype: str
         """
-        return self.get_instance(instance_id=instance_id).state["Name"]
+        return self.get_instances(instance_ids=[instance_id])[0][self.STATE][self.NAME]
```
If the instanced ID doesn't exist this probably returns a very unhelpful error (`list index out of range` for example.)

Please fix this, and and a unit test case covering this.

VijayantSoni replied:
`get_instances` would fail in that case with following error:
```
botocore.exceptions.ClientError: An error occurred (InvalidInstanceID.Malformed) when calling the DescribeInstances operation: Invalid id: "<instance_id>"
```

feluelle commented on airflow/providers/amazon/aws/hooks/ec2.py:
```diff
@@ -33,21 +33,90 @@ class EC2Hook(AwsBaseHook):
+        super().__init__(client_type="ec2", *args, **kwargs)
 
-    def get_instance(self, instance_id: str):
+    def stop_instances(self, instance_ids):
```
Please add type hints to all args and return values of your functions.

See https://docs.python.org/3/library/typing.html for more information.

feluelle commented on airflow/providers/amazon/aws/hooks/ec2.py:
```diff
@@ -33,21 +33,90 @@ class EC2Hook(AwsBaseHook):
+        :return: Response from EC2 describe_instances API
+        """
+        filters = [] if filters is None else filters
+        instance_ids = [] if instance_ids is None else instance_ids
```
```suggestion
        filters = filters or []
        instance_ids = instance_ids or []
```
See https://realpython.com/python-or-operator/#using-or-with-common-objects

feluelle replied:
@VijayantSoni can you fix this, please? :)

VijayantSoni committed:
rebasing with upstream master

RosterIn commented on airflow/providers/amazon/aws/hooks/ec2.py:
```diff
@@ -33,20 +33,82 @@ class EC2Hook(AwsBaseHook):
+        return [
+            instance
+            for reservation in description['Reservations']
+            for instance in reservation['Instances']
```
```suggestion
            instance for reservation in description['Reservations'] for instance in reservation['Instances']
```

RosterIn commented on airflow/providers/amazon/aws/hooks/ec2.py:
```diff
@@ -74,7 +136,11 @@ def wait_for_state(self, instance_id: str, target_state: str, check_interval: fl
         """
         instance_state = self.get_instance_state(instance_id=instance_id)
+
+        self.log.info("instance state: %s. Same as target: %s", instance_state, instance_state == target_state)
```
```suggestion
        self.log.info(
            "instance state: %s. Same as target: %s", instance_state, instance_state == target_state
        )
```

RosterIn commented on airflow/providers/amazon/aws/hooks/ec2.py:
```diff
@@ -74,7 +136,11 @@ def wait_for_state(self, instance_id: str, target_state: str, check_interval: fl
             time.sleep(check_interval)
             instance_state = self.get_instance_state(instance_id=instance_id)
+
+            self.log.info("instance state: %s. Same as target: %s", instance_state, instance_state == target_state)
```
```suggestion
            self.log.info(
                "instance state: %s. Same as target: %s", instance_state, instance_state == target_state
            )
```

RosterIn reviewed (changes requested):
The comments I left should fix the failing docs/pylint tests

VijayantSoni committed:
fix pylint issues

VijayantSoni committed:
use or instead of if-else

VijayantSoni committed:
Reapply changes lost after rebase strategy

VijayantSoni committed:
reapplying changes lost after rebase - II

ashb reviewed (dismissed):
Needs back-compat, and I'm not sure what real benefit this gives us.

I.e. why do we need this change?

Commit:
Added backward compatibility in EC2Hook for resource_type API
Added test cases for new methods added in EC2Hook
Restored previous test cases for EC2 hook and operator

Commit:
Fixing static code checks

Commit:
Fix static check in test_ec2.py

potiuk reviewed (approved):
LGTM. @ashb ?

ashb commented on airflow/providers/amazon/aws/hooks/ec2.py:
```diff
@@ -74,7 +192,15 @@ def wait_for_state(self, instance_id: str, target_state: str, check_interval: fl
         instance_state = self.get_instance_state(instance_id=instance_id)
+
+        self.log.info(
+            "instance state: %s. Same as target: %s", instance_state, instance_state == target_state
```
This log line is wrong -- we haven't checked it yet :)

VijayantSoni replied:
removed.

ashb reviewed (commented):
One log line to fix up

VijayantSoni committed:
remove log line
